docs/pipeline_semantics.md

```diff
@@ -0,0 +1,170 @@
+Conduit handled a message end-to-end and it can be safely discarded from the source.
+
+Pipeline nodes will either leave the message open and send it to the next node for processing or ack/nack it and not
+send it further down the pipeline. If the ack/nack fails, the node will stop running and return an error that will
+consequently stop the whole pipeline. The returned error is stored in the pipeline for further inspection by the user.
+
+### Message state change handlers
+
+A pipeline node can register state change handlers on a message that will be called when the message state changes. This
+is used for example to register handlers that reroute nacked messages to a dead letter queue or to update metrics when a
+message reaches the end of the pipeline.
+
+## Semantics
+
+### Messages are delivered in order
+
+Since messages are passed between nodes in channels and a node only processes one message at a time it is guaranteed
+that messages will flow through the Conduit pipeline in the same order as they are received by the source.
+
+There are two caveats:
+
+- If a pipeline contains multiple source connectors then the order of two messages coming from different connectors is
+  nondeterministic. Messages coming from the same source connector are still guaranteed to retain their order.
+- If a dead letter queue is configured then negatively acknowledged messages will be removed from the stream while the
+  pipeline will be kept running, thus impacting the order of messages.
+
+The order guarantee only holds inside of Conduit. Once a message reaches a destination connector it is allowed to buffer
+messages and batch write them to 3rd party systems. Normally the connector would retain the order, although we can't
+vouch for badly written connectors that don't follow this behavior.
+
+### Messages are delivered at least once
+
+Between pipeline restarts, it is guaranteed that any message that is processed successfully by all nodes and not
+filtered out will be delivered to a destination connector at least once. Multiple deliveries can occur in pipelines with
+multiple destinations that stopped because of a negatively acknowledged record, or pipelines, where a destination
+negatively acknowledged a record and processed more messages after that. For this reason, we strongly recommend
+implementing the write operation of a destination connector in an idempotent way (if possible).
+
+The delivery guarantee can be changed to "at most once" by adding a [dead letter queue](#Dead-letter-queue) handler that
+drops unsuccessfully processed messages.
+
+### Acks are delivered in order
+
+Conduit ensures that acknowledgments are sent to the source connector in the exact same order as records produced by the
+connector. This guarantee still holds, even if a destination connector acknowledges records in a different order or if a
+later record gets filtered out faster than an earlier record managed to get written to all destinations.
```

Comment:
I don't get this sentence.. could you elaborate more, please?

Reply:
In _Message states_ we explain the "Acked" status:

> Acked - a message was successfully processed and acknowledged. This can be done either by a processor (e.g. **it filtered the message out**) or by a destination.
> /.../
> Pipeline nodes will either leave the message open and send it to the next node for processing or **ack/nack it and not send it further down the pipeline**.

To sum up, "filtering a message" means acking it and not sending it to the next node.

Imagine a source connector that first produced message A and then message B. Message A would flow all the way to the destination, while message B would be filtered out by a processor. Since nodes are processing messages in parallel, message B could get filtered out (i.e. acked) faster than message A would reach the destination. Conduit still makes sure that acks will reach the source in the correct order - it waits for message A to be acked before sending the ack for message B back to the source, even though internally the acks might have happened in the opposite order.

The other scenario is that of a badly written destination connector that writes records asynchronously and messes up the order of acks sent back to Conduit - in that case Conduit would again make sure to correctly order the acks sent back to the source connector.

---

I'll see if I can reword it to make it clearer.

Reply:
@maha-hajja let me know if the rewording makes it clearer.

Reply:
Ah I had the whole meaning of "filtered out" misunderstood.. I thought filtered meant filtering some fields from the records and was wondering why would we ack it after that.
But now I get it! thank you

Reply:
@lovromazgon yep rewording is perfect